x-pack/plugin/voting-only-node/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/votingonly/VotingOnlyNodePluginTests.java

```diff
@@ -157,21 +155,21 @@ public void testVotingOnlyNodesCannotBeMasterWithoutFullMasterNodes() throws Exc
         internalCluster().setBootstrapMasterNodeIndex(0);
         internalCluster().startNode();
         internalCluster().startNodes(2, addRoles(Set.of(DiscoveryNodeRole.VOTING_ONLY_NODE_ROLE)));
-        internalCluster().startDataOnlyNodes(randomInt(2));
-        assertBusy(
-            () -> assertThat(
-                clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState().getLastCommittedConfiguration().getNodeIds().size(),
-                equalTo(3)
-            )
+        final int numDataNodes = randomInt(2);
+        internalCluster().startDataOnlyNodes(numDataNodes);
+        internalCluster().validateClusterFormed();
+
+        awaitClusterState(
+            state -> state.getLastCommittedConfiguration().getNodeIds().size() == 3 && state.nodes().size() == 3 + numDataNodes
         );
-        awaitMasterNode();
-        final String oldMasterId = clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState().nodes().getMasterNodeId();
+        final String oldMasterId = internalCluster().getMasterName();
 
         internalCluster().stopCurrentMasterNode();
         awaitMasterNotFound();
 
         // start a fresh full master node, which will be brought into the cluster as master by the voting-only nodes
         final String newMaster = internalCluster().startNode();
+        internalCluster().validateClusterFormed();
```

Comment:
I think we should revert this too, it's stronger than needed. `InternalTestCluster#getMasterName` already waits for the new master to be elected, and that's enough.